taskvine/src/manager/vine_manager.c

```diff
@@ -501,6 +505,167 @@ static int handle_transfer_hostport(struct vine_manager *q, struct vine_worker_i
+			break;
+		case (127):
+			warn(D_VINE,
+					"Task %d ran for a very short time and exited with code %d.\n",
+					t->task_id,
+					t->exit_code);
+			warn(D_VINE, "This usually means that the task's command could not be found, or that\n");
+			warn(D_VINE, "it uses a shared library not available at the worker, or that\n");
+			warn(D_VINE, "it uses a version of the glibc different than the one at the worker.\n");
+			break;
+		case (139):
+			warn(D_VINE,
+					"Task %d ran for a very short time and exited with code %d.\n",
+					t->task_id,
+					t->exit_code);
+			warn(D_VINE, "This usually means that the task's command had a segmentation fault,\n");
+			warn(D_VINE, "either because it has a memory access error (segfault), or because\n");
+			warn(D_VINE, "it uses a version of a shared library different from the one at the worker.\n");
+			break;
+		default:
+			break;
+		}
+	}
+
+	vine_task_info_add(q, t);
+
+	debug(D_VINE,
+			"%s (%s) done in %.02lfs total tasks %lld average %.02lfs",
+			w->hostname,
+			w->addrport,
+			(t->time_when_done - t->time_when_commit_start) / 1000000.0,
+			(long long)w->total_tasks_complete,
+			w->total_task_time / w->total_tasks_complete / 1000000.0);
+
+	return;
+}
+
+static vine_result_code_t get_completion_result(struct vine_manager *q, struct vine_worker_info *w, const char *line)
+{
+	if (!q || !w || !line)
+		return VINE_WORKER_FAILURE;
+
+	struct vine_task *t;
+	int task_status, exit_status;
+	uint64_t task_id;
+	int64_t output_length;
+
+	timestamp_t execution_time, start_time, end_time;
+	timestamp_t observed_execution_time;
+
+	// Format: task completion status, exit status (exit code or signal), output length, execution time, task_id
+	int n = sscanf(line,
+			"complete %d %d %" SCNd64 " %" SCNd64 " %" SCNd64 " %" SCNd64 "",
+			&task_status,
+			&exit_status,
+			&output_length,
+			&start_time,
+			&end_time,
+			&task_id);
+
+	if (n < 6) {
+		debug(D_VINE, "Invalid message from worker %s (%s): %s", w->hostname, w->addrport, line);
+		return VINE_WORKER_FAILURE;
+	}
+
+	execution_time = end_time - start_time;
+
+	/* If the worker sent back a task we have never heard of, then discard the following data. */
+	t = itable_lookup(w->current_tasks, task_id);
+	if (!t) {
+		debug(D_VINE,
+				"Unknown task completion from worker %s (%s): no task %" PRId64
+				" assigned to worker. Ignoring result.",
+				w->hostname,
+				w->addrport,
+				task_id);
+
+		time_t stoptime = time(0) + vine_manager_transfer_time(q, w, output_length);
+		link_soak(w->link, output_length, stoptime);
+		return VINE_SUCCESS;
+	}
+
+	if (task_status != VINE_RESULT_SUCCESS) {
+		w->last_failure_time = timestamp_get();
+		t->time_when_last_failure = w->last_failure_time;
+	}
+
+	/* If the task was forsaken by the worker or couldn't exeute, it didn't really complete, so short circuit. */
+	if (task_status == VINE_RESULT_FORSAKEN) {
+		t->forsaken_count++;
+		itable_remove(q->running_table, t->task_id);
```

Comment:
It looks like this `itable_remove`, `vine_task_set_result` and `change_task_state` are a common code path with the final few lines of this function.  Is there a way to simplify or combine here, so that we don't have duplicated paths through this function?  (I'm thinking about the stability of future edits to this code.)

Reply:
Moved calls to bottom of function